crates/wascc-provider/src/states/crash_loop_backoff.rs

```diff
@@ -1,20 +1,30 @@
-use kubelet::state::{State, Transition};
-use kubelet::{
-    pod::{Phase, Pod},
-    state,
-};
-
-use crate::{make_status, PodState};
+use crate::PodState;
+use kubelet::state::prelude::*;
 
 use super::registered::Registered;
 
-state!(
-    /// Pod has failed multiple times.
-    CrashLoopBackoff,
-    PodState,
-    {
+/// Pod has failed multiple times.
+#[derive(Default, Debug)]
+pub struct CrashLoopBackoff;
+
+#[async_trait::async_trait]
+impl State<PodState> for CrashLoopBackoff {
+    async fn next(
+        &self,
+        _pod_state: &mut PodState,
+        _pod: &Pod,
+    ) -> anyhow::Result<Transition<PodState>> {
         tokio::time::delay_for(std::time::Duration::from_secs(60)).await;
-        Ok(Transition::Advance(Box::new(Registered)))
-    },
-    { make_status(Phase::Pending, "CrashLoopBackoff") }
-);
+        Ok(Transition::next(self, Registered))
+    }
+
+    async fn json_status(
+        &self,
+        _pod_state: &mut PodState,
+        _pod: &Pod,
+    ) -> anyhow::Result<serde_json::Value> {
+        make_status(Phase::Pending, "CrashLoopBackoff")
+    }
+}
+
+impl EdgeTo<Registered> for CrashLoopBackoff {}
```

Comment:
I realise this may be one of those irreconcilable fundamental philosophical differences, but I want to make the case that the `EdgeTo` trait is not buying us anything in this particular approach to state machines.

The purpose of `EdgeTo` is for the compiler to prevent code making transitions that are not permitted by the state machine.  This is super useful when exposing a state machine where external code will drive the transitions.  If we are building an IoT coffee machine that users can program to make themselves the perfect brew, then we certainly don't want some unfortunate soul to put it into the Squirting state when it's not in the FullOfBrownStuff state.  (Look, don't @ me about the names, I don't know how coffee works.)

But that _isn't_ how we're doing it here.  In our approach, each state 'drives itself.'  The state machine switches based solely on code it controls.  _There's no other code we need to defend against._  The `EdgeTo` feature only defends the state against itself.

Meanwhile, there appears to be a downside to `EdgeTo` in which the `Transition::next` function ends up with a parameter that is _always_ ignored and only exists to carry type information a la phantom types.  We're cluttering the interface at the point of use, as well as requiring the busywork of declaring an edge as well as using it.

I realise that the 'defends the code against itself' feature does have merit; it's why we declare and type local variables.  But in this case I don't feel that explicitly specifying the subset of permissible transitions is worth the candle.  If it were driven from the outside, absolutely; but not when it's doing no more than checking the return value.

Thanks for coming to my TED talk.

Reply:
I agree that this is mostly a weak constraint. People are able to simply add edges in order to get their code to compile without considering the consequences, so for some developers we are just asking them to do extra typing (this actually feels a lot like `String` and `&str` for newbie Rust developers). I would also love to find a way to remove the unused parameter from `Transition::next`. But I do think that this gives people the option (and encourages them) to think about what transitions they are defining. It helps maintainers be confident that pull requests aren't introducing new state transitions without explicitly stating that. Personally, I think good Rust code is fairly verbose and explicit, so this matches my expectation of how one would define a state machine. You mention not needing to defend against outside code, and I disagree, having multiple authors contribute code (or even the same other on a different day) results in situations where they don't fully consider the behavior or intent of the rest of the code, and having the compiler there to double check things is one of the things I love about Rust. 

I'm happy to iterate on this API but in general I think there should be a limited set of valid outgoing edges associated with a state that is not simply every other state in the graph.

Reply:
I'll just add that I think there definitely are invalid state transitions in the graph, and if we can enforce them at compile time we should. Beyond that I think its just a question of ergonomics.

Reply:
Side note: could the `next` function use a `PhantomData`?

Reply:
@kflansburg I am not sure I understand your counterexample. Could you elaborate more? I don't see how `EdgeTo` would protect people contributing code on the same day in the way you explain. For example if both of us added a new state that transitions to starting, that would be caught by the git diff and probably the compiler as I'd probably be returning an incorrect type.

Either way, this is where I agree with @itowlson. If it is only protecting us from very specific edge cases, it does feel like the ceremony to benefit ratio is a little off, especially how adding a new state probably isn't something that happens very often.

Reply:
Yes, I think `PhantomData` would work here. 

I guess my point is that there are definitely invalid state transitions. For instance `Completed` to `Registered`. As it stands without edge enforcement, you could accidentally call `Transition::next(Registered)`, and you would be relying on your own understanding of the state machine and the code reviewers being vigilant to note that this is a transition which is invalid and has not been considered. I think this is a fairly real possibility, just based on our confusion over state names in the Wasi pull request. If we can make the compiler fail on this, requiring the developer and reviewers to explicitly add the edge, I think we should, and I dont think it is much more verbose (I actually think its helpful having a nice list of outgoing edges in the file). Without this enforcement I think we are leaving a lot of the benefits of a state machine on the table. 

At the very least, I would like to request that this is a feature that Providers can enable so that I can use it in my own CRI provider.